### PR TITLE
Handle all isolation checking for function calls in one place

### PR DESCRIPTION
**Pull request by DougGregor**
Isolation checking for calls had two separate implementation places:
one that looked at the declaration being called (for member
declarations) and one that worked on the actual call expression. Unify
on the latter implementation, which is more general and has access to
the specific call arguments. Improve diagnostics here somewher so we
don't regress in that area.

This refactoring shouldn't change the actual semantics, but it makes
upcoming semantic changes easier.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 1]
```diff
@@ -2646,6 +2596,8 @@ namespace {
         }
       }
 
+      // FIXME: Subscript?
```
Comment (ktoso) [inline lib/Sema/TypeCheckConcurrency.cpp:2599]: we don't support distributed subscripts (thankfully), so can remove the fixme I think?

Comment (ktoso) [reply lib/Sema/TypeCheckConcurrency.cpp:2599]: Looking at the line below I think we just handle it via fallthrough into the general case:

```swift
// This is either non-distributed variable, subscript, or something else.
       ctx.Diags.diagnose(declLoc,
                          diag::distributed_actor_isolated_non_self_reference,
```

so nothing to do here for the fixme?

Comment (DougGregor) [reply lib/Sema/TypeCheckConcurrency.cpp:2599]: I didn't change anything with respect to distributed subscripts, but wanted to note that this is where we'd deal with them if/when we do

File: test/Distributed/distributed_actor_isolation.swift [hunk 2]
```diff
@@ -207,10 +207,10 @@ distributed actor DijonMustard {
 
   convenience init(conv: FakeActorSystem) { // expected-warning {{initializers in actors are not marked with 'convenience'; this is an error in Swift 6}}{{3-15=}}
     self.init(system: conv)
-    self.f() // expected-error {{actor-isolated instance method 'f()' can not be referenced from a non-isolated context}}
+    self.f() // expected-error {{call to actor-isolated instance method 'f()' in a synchronous nonisolated context}}
   }
 
-  func f() {} // expected-note {{distributed actor-isolated instance method 'f()' declared here}}
+  func f() {} // expected-note {{calls to instance method 'f()' from outside of its actor context are implicitly asynchronous}}
```
Comment (ktoso) [inline test/Distributed/distributed_actor_isolation.swift:213]: yeah this message is good, the "more general" isolation rules just apply here 👍

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 3]
```diff
@@ -2434,41 +2415,6 @@ namespace {
       return ReferencedActor(var, isPotentiallyIsolated, ReferencedActor::NonIsolatedParameter);
     }
 
-    VarDecl *findReferencedBaseSelf(Expr *expr) {
```
Comment (ktoso) [inline lib/Sema/TypeCheckConcurrency.cpp:2437]: oh sweet to get rid of this entire func, thanks!

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 4]
```diff
@@ -2837,17 +2779,31 @@ namespace {
...
+
+        if (!fnType->getExtInfo().isAsync())
+          callOptions |= ActorReferenceResult::Flags::AsyncPromotion;
+        mayExitToNonisolated = false;
+
         break;
       }
 
+      // If we're calling an async function that's nonisolated, and we're in
+      // an isolated context, then we're exiting the actor context.
```
Comment (ktoso) [inline lib/Sema/TypeCheckConcurrency.cpp:2791]: 👍 that's not something we were diagnosing before right? Nice.

**Review by ktoso — APPROVED** [review]
AFAICS this looks good!

**Pull request merged**